File: php/Cpp/PhpLexerBase.cpp

#include "PhpLexerBase.h"
#include "PhpLexer.h"
#include "ystd.hpp"

using namespace antlr4;

using namespace std::string_view_literals;

std::unique_ptr<antlr4::Token> PhpLexerBase::nextToken()
{
    std::unique_ptr<antlr4::CommonToken> token = ystd::static_pointer_cast<antlr4::CommonToken>(antlr4::Lexer::nextToken());

    auto tokenType = token->getType();
    if (tokenType == PhpLexer::PHPEnd || tokenType == PhpLexer::PHPEndSingleLineComment)
    {
        if (this->mode == PhpLexer::SingleLineCommentMode)
        {
            // SingleLineCommentMode for such allowed syntax:
            // <?php echo "Hello world"; // comment ?>
            popMode(); // exit from SingleLineComment mode.
        }
        popMode(); // exit from PHP mode.

        if (token->getText() == "</script>"sv)
        {
            _phpScript = false;
            token->setType(PhpLexer::HtmlScriptClose);
        }
        else
        {
            // Add semicolon to the end of statement if it is absente.
            // For example: <?php echo "Hello world" ?>
            if (_prevTokenType == PhpLexer::SemiColon || _prevTokenType == PhpLexer::Colon
                || _prevTokenType == PhpLexer::OpenCurlyBracket || _prevTokenType == PhpLexer::CloseCurlyBracket)
            {
                token->setChannel(PhpLexer::SkipChannel);
            }
            else
            {
                token->setType(PhpLexer::SemiColon);
            }
        }
    }
    else if (tokenType == PhpLexer::HtmlName)
    {
        _htmlNameText = token->getText();
    }
    else if (tokenType == PhpLexer::HtmlDoubleQuoteString)
    {
        if (ystd::ic::iequals(token->getText(), "php"sv) &&
            _htmlNameText == "language"sv)
        {
            _phpScript = true;
        }
    }
    else if (this->mode == PhpLexer::HereDoc)
    {
        // Heredoc and Nowdoc syntax support: http://php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc
        switch (tokenType)
        {
        case PhpLexer::StartHereDoc:
        case PhpLexer::StartNowDoc:
            (void)0;
            {
                auto text = token->getText();
                text.erase(0, 3);
                _heredocIdentifier = std::move(ystd::replace(ystd::trim(text), "\'"sv, ""sv)); // token.Text.Substring(3).Trim().Trim('\'');
                break;
            }

        case PhpLexer::HereDocText:
            if (CheckHeredocEnd(token->getText()))
            {
                popMode();
                auto text = token->getText();
                ystd::trim(text);
                auto heredocIdentifier = GetHeredocIdentifier(text);
                if (ystd::ends_with(text, ";"sv))
                {
                    token->setText(heredocIdentifier += ";\n"sv);
                    token->setType(PhpLexer::SemiColon);
                }
                else
                {
                    token = ystd::static_pointer_cast<antlr4::CommonToken>(antlr4::Lexer::nextToken());
                    token->setText(heredocIdentifier += "\n;"sv);
                }
            }
            break;
        }
    }
    else if (this->mode == PhpLexer::PHP)
    {
        if (this->channel != HIDDEN)
        {
            _prevTokenType = tokenType;
        }
    }

    return token;
}

void PhpLexerBase::PushModeOnHtmlClose()
{
    popMode();
    if (_scriptTag)
    {
        if (!_phpScript)
        {
            pushMode(PhpLexer::SCRIPT);
        }
        else
        {
            pushMode(PhpLexer::PHP);
        }
        _scriptTag = false;
    }
    else if (_styleTag)
    {
        pushMode(PhpLexer::STYLE);
        _styleTag = false;
    }
}

void PhpLexerBase::PopModeOnCurlyBracketClose()
{
    if (_insideString)
    {
        _insideString = false;
        this->channel = PhpLexer::SkipChannel;
        popMode();
    }
}

File: php/Antlr4ng/PhpLexerBase.ts
import {CommonToken, Lexer, Token, CharStream} from "antlr4ng";
import { PhpParser } from "./PhpParser.js";
import { PhpLexer } from "./PhpLexer.js";

export default abstract class PhpLexerBase extends Lexer {
    private AspTags: boolean;
    protected _scriptTag: boolean;
    protected _styleTag: boolean;
    private _heredocIdentifier: string | undefined;
    private _prevTokenType: number;
    private _htmlNameText: string | undefined;
    private _phpScript: boolean;
    private _insideString: boolean;

    protected static MIN_CHAR_VALUE = 0x0000;
    protected static MAX_CHAR_VALUE = 0x10FFFF;

    constructor(input: CharStream) {
        super(input);
        this.AspTags = true;
        this._scriptTag = false;
        this._styleTag = false;
        this._heredocIdentifier = undefined;
        this._prevTokenType = 0;
        this._htmlNameText = undefined;
        this._phpScript = false;
        this._insideString = false;
    }

    nextToken() {
        let token = super.nextToken()

        if (token.type === PhpParser.PHPEnd || token.type === PhpLexer.PHPEndSingleLineComment) {
            if (this.mode === PhpLexer.SingleLineCommentMode) {
                // SingleLineCommentMode for such allowed syntax:
                // // <?php echo "Hello world"; // comment ?>
                this.popMode();
            }
            this.popMode();

            if (token.text === "</script>") {
                this._phpScript = false;
                token.type = PhpLexer.HtmlScriptClose;
            } else {
                // Add semicolon to the end of statement if it is absent.
                // For example: <?php echo "Hello world" ?>
                if (this._prevTokenType === PhpLexer.SemiColon || this._prevTokenType === PhpLexer.Colon || this._prevTokenType === PhpLexer.OpenCurlyBracket || this._prevTokenType === PhpLexer.CloseCurlyBracket) {
                    token.channel = 4; // Damn tool does not generate constants for declared channels.
                } else {
                    token.type = PhpLexer.SemiColon;
                }
            }
        }

        else if (token.type === PhpLexer.HtmlName) {
            this._htmlNameText = token.text
        }

        else if (token.type === PhpLexer.HtmlDoubleQuoteString) {
            if (token.text === "php" && this._htmlNameText === "language") {
                this._phpScript = true;
            }
        }

        else if (this.mode === PhpLexer.HereDoc) {
            // Heredoc and Nowdoc syntax support: http://php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc
            if (token.type === PhpLexer.StartHereDoc || token.type === PhpLexer.StartNowDoc) {
                this._heredocIdentifier = token.text.slice(3).trim().replace(/\'/, '').replace(/\'/, '');
            }

            else if (token.type === PhpLexer.HereDocText) {
                if (this.CheckHeredocEnd(token.text)) {
                    this.popMode()
                    const heredocIdentifier = this.GetHeredocIdentifier(token.text)
                    if (token.text.trim().endsWith(';')) {
                        token.text = `${heredocIdentifier};\n`;
                        token.type = PhpLexer.SemiColon;
                    } else {
                        token = super.nextToken()
                        token.text = `${heredocIdentifier}\n;`;
                    }
                }
            }
        }

        else if (this.mode === PhpLexer.PHP) {
            if (!(this.channel === PhpLexer.HIDDEN)) {
                this._prevTokenType = token.type;
            }
        }

        return token;
    }

    GetHeredocIdentifier(text: string): string {
        return text.trim().replace(/\;$/, "");
    }

    CheckHeredocEnd(text: string): boolean {
        return this.GetHeredocIdentifier(text) === this._heredocIdentifier;
    }

    IsNewLineOrStart(pos: number): boolean {
        return this.inputStream.LA(pos) <= 0 || this.inputStream.LA(pos) == '\r'.charCodeAt(0) ||
            this.inputStream.LA(pos) == '\n'.charCodeAt(0)
    }

    PushModeOnHtmlClose() {
        this.popMode();
        if (this._scriptTag) {
            if (!this._phpScript) {
                this.pushMode(PhpLexer.SCRIPT);
            } else {
                this.pushMode(PhpLexer.PHP);
            }
            this._scriptTag = false;
        } else if (this._styleTag) {
            this.pushMode(PhpLexer.STYLE);
            this._styleTag = false;
        }
    }

    HasAspTags(): boolean {
        return this.AspTags;
    }

    HasPhpScriptTag(): boolean {
        return this._phpScript;
    }

    PopModeOnCurlyBracketClose() {
        if (this._insideString) {
            this._insideString = false;
	    this.channel = 4; // Tool does not generate a constant for declared channels.
            this.popMode();
        }
    }

    ShouldPushHereDocMode(pos: number): boolean {
        return this.inputStream.LA(pos) === '\r'.charCodeAt(0) || this.inputStream.LA(pos) === '\n'.charCodeAt(0);
    }

    IsCurlyDollar(pos: number): boolean {
        return this.inputStream.LA(pos) === '$'.charCodeAt(0);
    }

    SetInsideString() {
        this._insideString = true
    }
}
